### PR TITLE
8247732: validate user-input intrinsic_ids in ControlIntrinsic

### PR DESCRIPTION
**Pull request by navyxliu**
/cc hotspot-compiler

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8247732](https://bugs.openjdk.java.net/browse/JDK-8247732): validate user-input intrinsic_ids in ControlIntrinsic

### Reviewers
 * [Nils Eliasson](https://openjdk.java.net/census#neliasso) (@neliasso - **Reviewer**) ⚠️ Review applies to 86cf452c067b64a717ae3d4d9434460666731b57
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1179/head:pull/1179`
`$ git checkout pull/1179`

### CODE REVIEW

**Comment by navyxliu** [comment]
/summary renew webrev to the latest jdk. fixed a typo and a bug.  Add constraints for both DisableIntrinsic and ControlIntrinsics.  Add tests to cover different use cases of them.

**Comment by navyxliu** [comment]
Here is the conversation we had a couple months ago. 
https://mail.openjdk.java.net/pipermail/hotspot-compiler-dev/2020-August/039742.html

The current behavior conforms to the table that Nils gave to me.

**Review by neliasso — COMMENTED** [review]
In general - I like it a lot. I will take it for a spin and go through the PR.

A heads up is that this PR will clash a bit with https://github.com/openjdk/jdk/pull/1276 which adds validation of the compile commands. Your change to compilerOracle.*pp seems well contained so there should be no problem to merge them.

**Comment by navyxliu** [comment]
Hi, Nils, 

Thank you for reviewing the lengthy PR!
The major part is to extend the testing framework to include ControlIntrinsic.  It may be also useful to test other compiler directives whose arguments are ccstr/ccstrlist.  

I am watching JDK-8256508.  I will update this PR accordingly if it needs. 

thanks, 
--lx

File: src/hotspot/share/compiler/compilerDirectives.hpp [hunk 1]
```diff
@@ -185,6 +185,37 @@ class ControlIntrinsicIter {
...
+ private:
+  bool _valid;
+  char* _bad;
+
+ public:
+  ControlIntrinsicValidator(ccstrlist option, bool disabled_all) : _valid(true), _bad(NULL) {
+    for (ControlIntrinsicIter iter(option, disabled_all); *iter != NULL && _valid; ++iter) {
+      if (vmIntrinsics::_none == vmIntrinsics::find_id(*iter)) {
+        _bad = NEW_C_HEAP_ARRAY(char, strlen(*iter) + 1, mtCompiler);
+        strncpy(_bad, *iter, strlen(*iter) + 1);
```
Comment (neliasso) [inline src/hotspot/share/compiler/compilerDirectives.hpp:198]: This doesn't compile. Using strlen as an argument to strncpy is disallowed.

> "warning: 'char* __builtin_strncpy(char*, const char*, long unsigned int)' specified bound depends on the length of the source argument [-Wstringop-overflow=]"

Do a min between strlen and the maximum allowed length.

Fix this for both uses of the string length (row 197 and 198).

Comment (navyxliu) [reply src/hotspot/share/compiler/compilerDirectives.hpp:198]: Out of curiosity,   what kind of gcc do you use?  I am using gcc/g++-8.4.0 and I do append `--with-extra-cxxflags='-Wstringop-overflow -Wstringop-truncation'`, why can't I trigger this warning? 

I read this. https://developers.redhat.com/blog/2018/05/24/detecting-string-truncation-with-gcc-8/
let me try to fix it by replacing strncpy with strcpy.

Comment (navyxliu) [reply src/hotspot/share/compiler/compilerDirectives.hpp]: hi, @neliasso, 

I update my toolchain from g++-8 to g++-10 and successfully reproduce this issue. 
I have updated my code to fix it. Actually, only one line is modified. 
```
@@ -195,7 +195,7 @@ class ControlIntrinsicValidator {
     for (ControlIntrinsicIter iter(option, disabled_all); *iter != NULL && _valid; ++iter) {
       if (vmIntrinsics::_none == vmIntrinsics::find_id(*iter)) {
         _bad = NEW_C_HEAP_ARRAY(char, strlen(*iter) + 1, mtCompiler);
-        strncpy(_bad, *iter, strlen(*iter) + 1);
+        strcpy(_bad, *iter);
         _valid = false;
       }
     }
```

The change has passed all regression test. I don't know why the table above doesn't update. 
https://github.com/navyxliu/jdk/actions/runs/375913238

Could you take a look my patch, again?

**Comment by navyxliu** [comment]
to adapt JDK-8256508,  I change compilerOracle.cpp as follows. 
I also merge 2 if constructs into one. 

```
      if (option == CompileCommand::ControlIntrinsic || option == CompileCommand::DisableIntrinsic) {
        ControlIntrinsicValidator validator(value, (option == CompileCommand::DisableIntrinsic));

        if (!validator.is_valid()) {
          jio_snprintf(errorbuf, buf_size, "Unrecognized intrinsic detected in %s: %s", option2name(option), validator.what());
        }
      }

```

File: src/hotspot/share/compiler/compilerDirectives.hpp [hunk 2]
```diff
@@ -185,6 +185,37 @@ class ControlIntrinsicIter {
...
+class ControlIntrinsicValidator {
+ private:
+  bool _valid;
+  char* _bad;
+
+ public:
+  ControlIntrinsicValidator(ccstrlist option, bool disabled_all) : _valid(true), _bad(NULL) {
+    for (ControlIntrinsicIter iter(option, disabled_all); *iter != NULL && _valid; ++iter) {
+      if (vmIntrinsics::_none == vmIntrinsics::find_id(*iter)) {
+        _bad = NEW_C_HEAP_ARRAY(char, strlen(*iter) + 1, mtCompiler);
```
Comment (neliasso) [inline src/hotspot/share/compiler/compilerDirectives.hpp:197]: Hi, 

I wanted something like this:

        size_t len = MIN(strlen(*iter), 64);        // cap len to a value we know is enough for all intrinsic names
        _bad = NEW_C_HEAP_ARRAY(char, len + 1, mtCompiler);
        strncpy(_bad, *iter, len + 1);                 // use strnlen to cap copy length

Comment (navyxliu) [reply src/hotspot/share/compiler/compilerDirectives.hpp]: okay. I see.  I make change to ensure the _bad is a bound string. here is a test. the _bad is limited to 64  even user-input is overflown. 
```
java -XX:+UnlockDiagnosticVMOptions -XX:ControlIntrinsic=+_dtan,+_0123456789012345678901234567890123456789012345678901234567890123456789 -version
Unrecognized intrinsic detected in ControlIntrinsic: _012345678901234567890123456789012345678901234567890123456789012
```

Comment (navyxliu) [reply src/hotspot/share/compiler/compilerDirectives.hpp]: limited the bound string 64 bytes instead of 65 in case it takes another cacheline.

**Review by neliasso — APPROVED** [review]
Looks good!

**Review by TobiHartmann — CHANGES_REQUESTED** [review]
This patch causes failures with `compiler/compilercontrol/commands/ControlIntrinsicTest.java`:
``` 
java.lang.RuntimeException: CompileCommand: option *Internal*.* missing in output: expected true, was false
	at jdk.test.lib.Asserts.fail(Asserts.java:594)
	at jdk.test.lib.Asserts.assertTrue(Asserts.java:486)
	at compiler.compilercontrol.share.processors.CommandProcessor.check(CommandProcessor.java:73)
```

**Comment by TobiHartmann** [comment]
`compiler/compilercontrol/mixed/RandomValidCommandsTest.java` fails as well:
```
java.lang.RuntimeException: CompileCommand: option *.*$method missing in output: expected true, was false
	at jdk.test.lib.Asserts.fail(Asserts.java:594)
	at jdk.test.lib.Asserts.assertTrue(Asserts.java:486)
	at compiler.compilercontrol.share.processors.CommandProcessor.check(CommandProcessor.java:73)
```

**Comment by navyxliu** [comment]
take a look the failure.

**Comment by navyxliu** [comment]
**Analysis**

The failure of compiler/compilercontrol/commands/ControlIntrinsicTest.java is reproducible with `-XX:CompileThreshold=100`. RandomValidCommandsTest.java is a flaky test. Once it generates INTRINSIC commands in random test, it fails due to the same reason as  the first case. 

After JDK-8256508, the output message of option has changed. 
now it outputs  message such as `CompileCommand: ControlIntrinsic *Klass.-()V const char* ControlIntrinsic = '+_newArray -_minF +_copyOf'` even thought the input  is `-XX:Compilecommand=option,*Klass.-()V,ControlIntrinsic,ccstr,+_newArray,-_minF,+_copyOf`. 

**Solution**
I need to reformat CommandProcessor.getOutputString in test/hotspot/jtreg/compiler/compilercontrol/share/processors/CommandProcessor.java

**Comment by navyxliu** [comment]
Random  test `compiler/compilercontrol/mixed/RandomCommandsTest.java` may generate invalid method patterns and then run into [JDK-8257800](https://bugs.openjdk.java.net/browse/JDK-8257800). I need to solve that bug first.

File: src/hotspot/share/compiler/compilerDirectives.hpp [hunk 3]
```diff
@@ -185,6 +185,39 @@ class ControlIntrinsicIter {
...
+  bool _valid;
+  char* _bad;
+
+ public:
+  ControlIntrinsicValidator(ccstrlist option, bool disabled_all) : _valid(true), _bad(nullptr) {
+    for (ControlIntrinsicIter iter(option, disabled_all); *iter != NULL && _valid; ++iter) {
+      if (vmIntrinsics::_none == vmIntrinsics::find_id(*iter)) {
+        const size_t len = MIN2<size_t>(strlen(*iter), 63) + 1;  // cap len to a value we know is enough for all intrinsic names
+        _bad = NEW_C_HEAP_ARRAY(char, len, mtCompiler);
+        // strncpy always write len characters.if the source string is shorter, the function fills the remaining bytes with NULs.
```
Comment (TobiHartmann) [inline src/hotspot/share/compiler/compilerDirectives.hpp:199]: Some typos, should be: `// strncpy always writes len characters. If the source string is shorter, the function fills the remaining bytes with NULLs.`

File: src/hotspot/share/compiler/compilerDirectives.hpp [hunk 4]
```diff
@@ -185,6 +185,39 @@ class ControlIntrinsicIter {
   ControlIntrinsicIter& operator++();
 };
 
+class ControlIntrinsicValidator {
+ private:
+  bool _valid;
+  char* _bad;
+
+ public:
+  ControlIntrinsicValidator(ccstrlist option, bool disabled_all) : _valid(true), _bad(nullptr) {
```
Comment (TobiHartmann) [inline src/hotspot/share/compiler/compilerDirectives.hpp:194]: The argument of `ControlIntrinsicIter` is called `disable_all`, so it should be `disable_all` here as well.

The comment in `ControlIntrinsicIter` is confusing:
```
// if disable_all is true, it accepts DisableIntrinsic(deprecated) and all intrinsics
// appear in the list are to disable
```

Could you explain what exactly `disabled_all` is needed for?

File: src/hotspot/share/compiler/directivesParser.cpp [hunk 5]
```diff
@@ -319,6 +319,23 @@ bool DirectivesParser::set_option_flag(JSON_TYPE t, JSON_VAL* v, const key* opti
...
+
+        if (strncmp(option_key->name, "ControlIntrinsic", 16) == 0) {
+          ControlIntrinsicValidator validator(s, false/*disabled_all*/);
+
+          if (!validator.is_valid()) {
+            error(VALUE_ERROR, "Unrecognized intrinsic detected in ControlIntrinsic: %s", validator.what());
+            return false;
+          }
+        }
+        else if (strncmp(option_key->name, "DisableIntrinsic", 16) == 0) {
```
Comment (TobiHartmann) [inline src/hotspot/share/compiler/directivesParser.cpp:331]: Please change to `} else if (...) {`.

**Review by TobiHartmann — CHANGES_REQUESTED** [review]

File: src/hotspot/share/compiler/compilerDirectives.hpp (hunk 4 above)
Comment (navyxliu) [reply src/hotspot/share/compiler/compilerDirectives.hpp:194]: before ControlIntrinsic, we have only this option: 
-XX:DisableIntrinsic=_id1,_id2,_id3,...

With ControlIntrinsic, we can have this form, + means enable, - means disable. 
-XX:ControlIntrinsic=+_id1, -_id2,+_id3,...

After DisableIntrinsic, we can still use this form. in this case, ControlIntrinsic is just an alias of  DisableIntrinsic. 
-XX:ControlIntrinsic=_id1, _id2,_id3,...

**Comment by navyxliu** [comment]
[ping] Could reviewer take a look at this PR?

**Comment by navyxliu** [comment]
hi, @TobiHartmann, Could you take a look at this one?

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Comment by navyxliu** [comment]
I ran test-tier1 with JTREG="VM_OPTIONS=-XX:CompileThreshold=100" again. it's clear. 
I think it's safe. May I ask a sponsor to merge it?

**Comment by navyxliu** [comment]
Thanks you, all reviewers!

**Pull request closed without merging**